### PR TITLE
Handle system headers differently on the command line for GCC

### PR DESCRIPTION
**Pull request by ghale**
See https://github.com/gradle/gradle-native/issues/583

This adds support for specifying system header directories on the command line for GCC-compatible toolchains using "-isystem" rather than "-I"

### CODE REVIEW

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/AbstractNativeCompileSpec.java [hunk 1]
```diff
@@ -62,6 +63,21 @@ public void include(Iterable<File> includeRoots) {
...
+        return systemIncludeRoots;
+    }
+
+    @Override
+    public void systemInclude(Iterable<File> systemIncludeRoots) {
+        addAll(this.systemIncludeRoots, systemIncludeRoots);
+    }
+
+    @Override
+    public void systemInclude(File... systemIncludeRoots) {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/AbstractNativeCompileSpec.java]: Aren't we always going to use the `Iterable` version above?

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java [hunk 2]
```diff
@@ -243,6 +247,25 @@ public void includes(Object includeRoots) {
...
+    public ConfigurableFileCollection getSystemIncludes() {
+        return systemIncludes;
+    }
+
+    /**
+     * Add system include directories where the compiler should search for header files.
+     *
+     * @since 4.8
+     */
+    public void systemIncludes(Object systemIncludeRoots) {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java]: I think we can limit the API exposure to just `getSystemIncludes()`.  I think that makes it more "Provider" like.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java [hunk 3]
```diff
@@ -243,6 +247,25 @@ public void includes(Object includeRoots) {
         includes.from(includeRoots);
     }
 
+    /**
+     * Returns the system include directories to be used for compilation.
+     *
+     * @since 4.8
+     */
+    @Internal("The paths for include directories are tracked via the includePaths property, the contents are tracked via discovered inputs")
+    public ConfigurableFileCollection getSystemIncludes() {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java]: We should add this to the DSL docs too

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/CompileTaskConfig.java [hunk 4]
```diff
@@ -93,7 +93,7 @@ public FileCollection transform(NativeDependencySet original) {
             }
         });
         FileCollectionFactory fileCollectionFactory = ((ProjectInternal) task.getProject()).getServices().get(FileCollectionFactory.class);
-        task.includes(fileCollectionFactory.create(new MinimalFileSet() {
+        task.systemIncludes(fileCollectionFactory.create(new MinimalFileSet() {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/CompileTaskConfig.java]: We should add something about this to the release notes

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/AbstractNativeCompileSpec.java (hunk 1 above)
Comment (ghale) [reply subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/AbstractNativeCompileSpec.java]: Yeah, I was just making it consistent with `include()`.  I've removed it.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java (hunk 2 above)
Comment (ghale) [reply subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java]: Fair enough.  Removed.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java (hunk 3 above)
Comment (ghale) [reply subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java]: Done.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/CompileTaskConfig.java (hunk 4 above)
Comment (ghale) [reply subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/CompileTaskConfig.java]: Done.

**Comment by ghale** [comment]
@big-guy this is ready for another look when you get a chance.

**Pull request merged**